sdk/sync/subscription.go

```diff
@@ -52,39 +35,50 @@ func (s *subscription) process() {
 
 	startSeq, err := s.client.XLen(key).Result()
 	if err != nil {
-		s.connCh <- -1
-		s.result = fmt.Errorf("failed to fetch current length of stream: %w", err)
+		s.w.re.SLogger().Errorf("failed to fetch current length of stream: %w", err)
 		return
 	}
 
 	log := s.w.re.SLogger().With("subtree", s.subtree, "start_seq", startSeq)
 
-	// Get a connection and store its connection ID, so that stop() can unblock
-	// it upon closure.
+	// Get a connection and store its connection ID, so we can unblock it when canceling.
 	conn := s.client.Conn()
 	defer conn.Close()
 
-	id, err := conn.ClientID().Result()
-	if err != nil {
-		s.connCh <- -1
-		s.result = fmt.Errorf("failed to get the current conn id: %w", err)
-		return
-	}
+	connID, err := conn.ClientID().Result()
+	done := make(chan struct{})
+	closed := make(chan struct{})
+	go func() {
+		defer close(closed)
+		// we need a _non_ canceled client for this to work.
+		client := s.client.WithContext(context.Background())
+		select {
+		case <-s.client.Context().Done():
+			err := client.ClientUnblockWithError(connID).Err()
+			if err != nil {
+				log.Errorw("failed to kill connection", "error", err)
+			}
+		case <-done:
+			// no need to unblock anything.
+		}
+	}()
 
-	// store the conn ID in the channel.
-	s.connCh <- id
+	defer func() {
+		close(done)
+		<-closed
+	}()
```

Comment:
@nonsense thoughts on how to write this better? This works, it's just a little messy (cleaner than requiring a separate stop method, IMO, but still messy.